collapse.go

```diff
@@ -310,6 +316,46 @@ func (c *CollapseContext) processRewardWithdrawals(blockIndex uint64) {
 	c.rewardWithdrawalRequests = leftovers
 }
 
+func (c *CollapseContext) processFinalizedTransactions(height uint64, finalized []*Transaction) {
+	for _, tx := range finalized {
+		key := make([]byte, len(keyTransactionFinalized)+8+32)
+
+		copy(key[:len(keyTransactionFinalized)], keyTransactionFinalized[:])
+		binary.BigEndian.PutUint64(key[len(keyTransactionFinalized):len(keyTransactionFinalized)+8], height)
+		copy(key[len(keyTransactionFinalized)+8:len(keyTransactionFinalized)+8+32], tx.ID[:])
+
+		c.tree.Insert(key, nil)
+	}
+
+	pruningLimit := uint64(conf.GetPruningLimit())
+
+	if height < pruningLimit {
+		return
+	}
+
+	heightLimit := height - pruningLimit
+
+	cb := func(k, v []byte) bool {
+		r := bytes.NewReader(k)
+
+		var buf [8]byte
+		if n, err := r.Read(buf[:]); n != 8 || err != nil {
+			return false
+		}
+
+		height := binary.BigEndian.Uint64(buf[:])
+
+		if height <= heightLimit {
+			c.tree.Delete(append(keyTransactionFinalized[:], k...))
+			return true
+		}
+
+		return false
+	}
+
+	c.tree.IteratePrefix(keyTransactionFinalized[:], cb)
```

Comment:
Hm, that's only if we assume that those transactions will always be present in `(*Transaction).buffer`. Since we keep track of finalized transaction IDs in the ledger state, we should delete them separately upon `collapseTransactions`.

Reply:
they should be always present, right? if we'll have mismatch between transactions in buffer and avl thats a sign of some big problems